src/core/ddsc/src/dds_reader.c

```diff
@@ -127,13 +126,13 @@ void dds_reader_data_available_cb (struct dds_reader *rd)
 
       lst->on_data_on_readers (sub->m_hdllink.hdl, lst->on_data_on_readers_arg);
 
-      ddsrt_mutex_lock (&rd->m_entity.m_observers_lock);
       ddsrt_mutex_lock (&sub->m_observers_lock);
       sub->m_cb_count--;
       sub->m_cb_pending_count--;
       ddsrt_cond_broadcast (&sub->m_observers_cond);
     }
     ddsrt_mutex_unlock (&sub->m_observers_lock);
+    ddsrt_mutex_lock (&rd->m_entity.m_observers_lock);
```

Comment:
Looks fine to me. Probably the hierarchy is unimportant for *m_observers_lock*, but just wondering unlock of subscriber isn't better done after locking the reader? (the opposite would of course be true when locking the subscriber and unlocking the reader)

Reply:
You can go up the hierarchy, never down. (So the few functions that do need to go down, like pushing listeners or changing QoS jump through hoops to be able to resume iterating over the children after having unlocked the parent prior to locking a child.)

For once I did even add a comment 😀:
```C
  ddsrt_mutex_t m_observers_lock;   /* locking parent->...->m_observers_lock while holding it is allowed */
```

No state of the reader is touched between unlocking the reader's ``m_observers_lock`` and re-locking it. Indeed, the entire function is more-or-less a no-op if you end up in the "data on readers" case. Locking the reader after unlocking the subscribers is ok as far as mutual exclusion on the reader's internal state is concerned.

Then there is the ordering of callbacks. Listener invocations on a single entity have always been serialised. This happens through the ``m_cb_count`` field, which is itself protected by ``m_observers_lock``. The reader has incremented its ``m_cb_count`` field, therefore no other thread will make it past the barrier at the top of the function until the end up of this function.

Finally — for what I can think of, I may be mistaken — functions like setting listeners and status masks use the same counts to ensure they are serialised with respect to listener invocations, so such changes won't be happening either.

And thus, I can't think of observable behavioural differences depending on whether ``m_observers_lock`` is re-locked after unlocking the subscriber, or before re-locking the subscriber. And then I think the current structure is cleaner.